c3d/src/C3d.cpp

```diff
@@ -81,6 +92,79 @@ class C3dSimpleSpelling: public ParsedAttrInfo {
 private:
   Spelling S[1];
 
+protected:
+  static std::vector<c3d_var_info> c3d_scope;
+
+  static void C3dPushVar(Parser *P, SourceLocation StartLoc, SourceLocation IdLoc,
+                         IdentifierInfo *Id, QualType T,
+                         TypeSourceInfo *TInfo, StorageClass S) {
+      LLVM_DEBUG(llvm::dbgs() << "c3d pushing variable " << Id->getName() << " \n");
+      c3d_var_info v = {
+          .StartLoc  = StartLoc,
+          .IdLoc     = IdLoc,
+          .Id        = Id,
+          .T         = T,
+          .TInfo     = TInfo,
+          .S         = S,
+      };
+      c3d_scope.push_back(v);
+  }
+
+  static void C3dClearScope() {
+    LLVM_DEBUG(llvm::dbgs() << "c3d CLEARING scope\n");
+    c3d_scope.clear();
+  }
+
+  // Pop one variable from the C3d scope (FIXME: something more generic please)
+  static void C3dPopVar() {
+      c3d_scope.pop_back();
```

Comment:
but this doesn't remove it from the clang scope, right? so this is breaking the invariant that `c3d_scope` contains exactly the set of variables we added to the clang scope

Reply:
When we call this (currently), the clang scope is gone. They exist only for the brief moment of parsing an expression (see `C3dParseExpr`) and are immediately popped. This line is removing the `this` from the set of variables that we will add the _next_ time we make a clang scope.

I see your point in that
```cpp
C3dEnterScope();
C3dPopVar()
```
makes a weird state, maybe we can forbid that. Or, actually, just make `C3dEnterScope` private and provide structured wrappers for it (like `C3dParseExpr`).